packages/core/src/modules/job-listing/job-listing-service.ts

```diff
@@ -29,127 +36,67 @@ export class JobListingServiceImpl implements JobListingService {
-
-    const jobListing = await this.jobListingRepository.update(id, rest)
-
-    const allLocations = await this.jobListingLocationRepository.getAll()
-
-    const currentLocationIds = this.getCurrentLocationIds(jobListing, allLocations)
-    const newLocationNames = this.getNewLocationNames(locations, jobListing)
-    const newLocationIds = this.getNewLocationIds(newLocationNames, allLocations)
-
-    this.validateLocations(currentLocationIds, jobListing, newLocationIds, newLocationNames)
+  async updateJobListingById(id: JobListingId, { locations, ...input }: JobListingWrite): Promise<JobListing> {
+    const existing = await this.jobListingRepository.getById(id)
+    const merged: JobListingWrite = { ...existing, ...input, locations }
 
-    const { toRemove, toAdd } = this.determineLocationChanges(currentLocationIds, locations, allLocations)
-    await this.handleLocationChanges(toRemove, toAdd, jobListing, newLocationNames)
+    this.validateWriteModel(merged)
 
+    const jobListing = await this.jobListingRepository.updateJobListingById(id, input)
+    const diff = this.getLocationDiff(jobListing.locations, locations)
+    await this.applyLocationDiff(diff.toRemove, diff.toAdd, jobListing)
     return {
       ...jobListing,
       locations,
     }
   }
 
-  private getCurrentLocationIds(jobListing: JobListing, allLocations: LocationSelect[]): string[] {
-    return jobListing.locations
-      .map((loc) => allLocations.find((x) => x.name === loc))
-      .filter((x) => x !== undefined)
-      .map((x) => x?.id) as string[]
+  private validateWriteModel(input: JobListingWrite): void {
+    assert(isAfter(input.start, new Date()), new InvalidStartDateError("Start date cannot be before today"))
+    assert(isAfter(input.end, input.start), new InvalidEndDateError("End date cannot be before start date"))
+    assert(
+      input.deadline !== null ? isBefore(input.deadline, input.start) : true,
+      new InvalidDeadlineError("Deadline cannot be after start date")
+    )
+    assert(input.locations.length > 0, new InvalidLocationError("At least one location is required"))
   }
 
-  private getNewLocationNames(providedLocations: string[], jobListing: JobListing): string[] {
-    return providedLocations.filter((x) => !jobListing.locations.includes(x))
-  }
-
-  private getNewLocationIds(newLocationNames: string[], allLocations: LocationSelect[]): string[] {
-    return newLocationNames
-      .map((name) => allLocations.find((x) => x.name === name)?.id || null)
-      .filter((x) => x !== null) as string[]
-  }
-
-  private validateLocations(
-    currentLocationIds: string[],
-    jobListing: JobListing,
-    newLocationIds: string[],
-    newLocationNames: string[]
-  ) {
-    if (
-      currentLocationIds.length !== jobListing.locations.length ||
-      newLocationIds.length !== newLocationNames.length
-    ) {
-      throw new NotFoundError(`Could not update JobListing(${jobListing.id})`)
-    }
-  }
-
-  private determineLocationChanges(
```

Comment:
No longer needed, because we removeByLocationName instead of having to deal with the ids